### ISSUE TITLE
Allow image descriptions to be read automatically by screen readers without needing to expand media

### ISSUE THREAD

**Issue by PratikP1** [nolanlawson/pinafore#2257]
When a screen reader user traverses the timeline, alt texts/image descriptions are not automatically read by their screen readers. Regardless of whether there is alt text associated with the media, the user must expand the media before they can discover if there is no description associated with it. Additional action should not be required to discover alt texxt or reading it.

There are a few potential solutions:

1. Indicate to the user that there is alt text associated with the media without passing on the alt text.
2. Indicate that there is alt text and provide a keyboard command to have the description read.
3. Indicate that there is alt text and automatically read the description. 

My personal preference is for either 2 or 3 with 3 being the best option. If the user wishes to look at the description in detail, there is always the option of expanding the media. That should be the last resort.

**Comment by PratikP1** [comment]
Possibly related to #2250.

**Comment by nolanlawson** [comment]
Interesting suggestion. So here is how things currently work (after a quick glance):

1. The `aria-label` on an `<article>` does not include the alt text for any image, but it does say "has media"
2. For each `<img>` there is a corresponding `<button>` surrounding it, which you can press to pop open the modal
3. Both `<img>`s have alt text
4. The surrounding `<button>`s also have an `aria-label` saying "Show image: " plus the alt text

So it's not necessarily the case that you need to pop open the modal to read the alt text, but I guess depending on the screenreader/UA it may be difficult without popping open the modal?

My understanding of the spirit of @MarcoZehe's suggestion for `aria-label`s on the `<article>` is that it allows you to quickly read the timeline by pressing up/down (or j/k) without needing to focus on individual toot contents. In that spirit, it seems like maybe the alt text for all images should be concatenated to the `aria-label` on the `<article>`? Not sure if that makes the label a bit overwhelming though. It also gets complex with sensitive media, which ideally shouldn't have its alt text read by default.

**Comment by PratikP1** [comment]
@nolanlawson commented:

> So it's not necessarily the case that you need to pop open the modal to read the alt text, but I guess depending on the screenreader/UA it may be difficult without popping open the modal?

The only way to read  alt text without expanding the media is if you disable left/right keys, which--by default--move from timeline to timeline. Disabling the default behavior allows me to go through a post's blocks such as user name, hashtags, and the "show image" button. So far, I've tested with Firefox, and Chromium based browsers using NVDA as well as Voiceover on iOS. I don't have a working Mac configuration right now to test. The behavior is consistent. If I want to hear the description, I'd need to either expand the media or change the default behavior of left/right arrow keys then move to the "show image" button for each post. Once again, this requires multiple key presses.

> In that spirit, it seems like maybe the alt text for all images should be concatenated to the aria-label on the <article>? Not sure if that makes the label a bit overwhelming though. It also gets complex with sensitive media, which ideally shouldn't have its alt text read by default.

This seems like the best solution. As to the verboseness, Pinafore's speed is an advantage here. With all screen readers, the software is quite performant. Pressing either the control key or arrow keys to move to other posts interrupts speech quite nicely.

How does Pinafore handle sensitive image viewing right now? Is the function of noting that an image is sensitive controlled by Pinafore or Mastodon's API? I haven't encountered a sensitive image so far. What information gets conveyed to the screen reader user if they encounter a sensitive image? The solution to this thorny issue could lie in the answers to this question.

**Comment by nolanlawson** [comment]
> The only way to read alt text without expanding the media is if you disable left/right keys

Hmm, it should work if you press the <kbd>Tab</kbd> key.

> How does Pinafore handle sensitive image viewing right now?

IIRC it's just a button that says "click to show" and then when you show it, it shows the image with the alt.

> Is the function of noting that an image is sensitive controlled by Pinafore or Mastodon's API?

Mastodon's

> What information gets conveyed to the screen reader user if they encounter a sensitive image?

I believe it has a button that says "show sensitive" or something like that.

**Comment by jcsteh** [comment]
I think it would be reasonable to include the alt text for non-sensitive images in the article aria-label. As I understand it, for sensitive images, the alt text isn't shown unless you explicitly open the modal, so it also makes sense not to include it in the aria-label.
